### PR TITLE
executor: fix panic issue when handle `HashAggExec.Close()`

### LINKED ISSUE #41778: close of closed channel in (*HashAggExec).Close
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
CREATE TABLE `6ea6fc5a` (
                            `b28e1129` json DEFAULT NULL,
                            `e6f9d33b` decimal(31,1) DEFAULT '-229239137389541701499094015080.1'
);

CREATE TABLE `02ed3374` (
                            `1b60d2bf` smallint(6) DEFAULT '-11385',
                            `3a7ec6ea` decimal(63,14) DEFAULT '-6197127648752447138876497216172307937317445669286.98661563645110',
                            KEY `76114f1d` (`1b60d2bf`)
) ENGINE=InnoDB DEFAULT CHARSET=ascii COLLATE=ascii_bin COMMENT='be783ecc'
    PARTITION BY HASH (`1b60d2bf`) PARTITIONS 3;


CREATE TABLE `cead3486` (
                            `1c73b98b` text DEFAULT NULL,
                            `b4e81697` float NOT NULL DEFAULT '1.8132474',
                            PRIMARY KEY (`b4e81697`) /*T![clustered_index] CLUSTERED */
);


insert into `6ea6fc5a` values (NULL,   1228.0);
insert into `6ea6fc5a` values ('"json string1"' ,    623.8);
insert into `6ea6fc5a` values (NULL,   1337.0);
insert into `6ea6fc5a` values ('-44' ,     24.0);
insert into `6ea6fc5a` values ('"json string1"' , 882543.0);
insert into `6ea6fc5a` values ('"json string1"' ,     24.0);
insert into `6ea6fc5a` values ('323232323.32323235' ,     24.0);
insert into `6ea6fc5a` values ('-44' ,     24.0);
insert into `6ea6fc5a` values ('"json string1"' ,     24.0);
insert into `6ea6fc5a` values ('"json string1"' ,     24.0);

select count(*)
from `6ea6fc5a`
where not (`6ea6fc5a`.`b28e1129` in (select `3a7ec6ea`
                                     from `02ed3374`
                                     where not (`6ea6fc5a`.`b28e1129` in (select `1c73b98b`
                                                                          from `cead3486`
                                                                          where `6ea6fc5a`.`e6f9d33b` in
                                                                                (select `3a7ec6ea`
                                                                                 from `02ed3374`
                                                                                 where not (`6ea6fc5a`.`e6f9d33b` between 61151 and 260381)
                                                                                   and IsNull(`6ea6fc5a`.`b28e1129`))))
                                        ));
```

### PR DESCRIPTION
**Pull request by solotzg**
### What problem does this PR solve?

Issue Number: close #41778

Problem Summary:

`func (e *HashAggExec) Close() error` may be invoked multiple times and this part should guarantee idempotent
https://github.com/pingcap/tidb/blob/6f1bdf76cf3e1ea0496ca6217b8afdad745a1b5d/executor/aggregate.go#L267-L291

### What is changed and how it works?

- rename `parallelExecInitialized` to `parallelExecValid`
- set `parallelExecValid` to `false` after relevant channel closed.

The result becomes like `ERROR 1815 (HY000): expression isnull(cast(test.ta.a1, var_string(4294967295))) cannot be pushed down` because `func (pc *PbConverter) encodeDatum`  has not supported json

https://github.com/pingcap/tidb/blob/6f1bdf76cf3e1ea0496ca6217b8afdad745a1b5d/expression/expr_to_pb.go#L97-L155

### Tests

Reproduction:

```
CREATE database if not exists test;
use test;
CREATE TABLE `ta` (
    `a1` json DEFAULT NULL,
    `a2` decimal(31, 1) DEFAULT '-229239137389541701499094015080.1'
);
CREATE TABLE `tb` (
    `b1` smallint(6) DEFAULT '-11385',
    `b2` decimal(63, 14) DEFAULT '-6197127648752447138876497216172307937317445669286.98661563645110',
    KEY `76114f1d` (`b1`)
) ENGINE = InnoDB DEFAULT CHARSET = ascii COLLATE = ascii_bin COMMENT = 'be783ecc' PARTITION BY HASH (`b1`) PARTITIONS 3;
CREATE TABLE `tc` (
    `c1` text DEFAULT NULL,
    `c2` float NOT NULL DEFAULT '1.8132474',
    PRIMARY KEY (`c2`)
    /*T![clustered_index] CLUSTERED */
);
insert into `ta`
values (NULL, 1228.0);
insert into `ta`
values ('"json string1"', 623.8);
insert into `ta`
values (NULL, 1337.0);
insert into `ta`
values ('-44', 24.0);
insert into `ta`
values ('"json string1"', 882543.0);
insert into `ta`
values ('"json string1"', 24.0);
insert into `ta`
values ('323232323.32323235', 24.0);
insert into `ta`
values ('-44', 24.0);
insert into `ta`
values ('"json string1"', 24.0);
insert into `ta`
values ('"json string1"', 24.0);
select count(*)
from `ta`
where not (
        `ta`.`a1` in (
            select `b2`
            from `tb`
            where not (
                    `ta`.`a1` in (
                        select `c1`
                        from `tc`
                        where `ta`.`a2` in (
                                select `b2`
                                from `tb`
                                where not (
                                        `ta`.`a2` between 61151 and 260381
                                    )
                                    and IsNull(`ta`.`a1`)
                            )
                    )
                )
        )
    );
```

Result:

```
mysql> select count(*) from `ta` where not (         `ta`.`a1` in (             select `b2`             from `tb`             where not (
     `ta`.`a1` in (                         select `c1`                         from `tc`                         where `ta`.`a2` in (
              select `b2`                                 from `tb`                                 where not (                                         `ta`.`a2` between 61151 and 260381                                     )                                     and IsNull(`ta`.`a1`)
     )                     )                 )         )     );
ERROR 1815 (HY000): expression isnull(cast(test.ta.a1, var_string(4294967295))) cannot be pushed down
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/aggregate/agg_hash_executor.go [hunk 1]
```diff
@@ -327,7 +328,7 @@ func (e *HashAggExec) initForParallelExec(_ sessionctx.Context) {
 		e.finalWorkers[i].finalResultHolderCh <- exec.NewFirstChunk(e)
 	}
 
-	e.parallelExecInitialized = true
```
Comment (hawkingrei) [inline executor/aggregate/agg_hash_executor.go:330]: Can you write a test case for it?

Comment (solotzg) [reply executor/aggregate/agg_hash_executor.go:330]: done

**Review by wshwsh12 — APPROVED** [review]

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**